### ISSUE TITLE
[$300] Investigate text / flair for copilots / admins

### ISSUE THREAD

**Issue by jmgasper** [topcoder-platform/forums-plugins#27]
@atelomycterus - One thing that Topcoder would like to see is the ability to show flair or additional text next to admin and copilot handles, based on their role from the user object.

The question that Topcoder posed:

`can we add the roles in forum? like copilot, members, admin, client in the challenge forum some way to indicate the role the person has in the challenge?`

Note that this would be *only* for challenge forums inside a group that the copilot is actively managing, which could be a bit messy, since we'll need to add in dependencies on the challenge and resource APIs

* https://github.com/topcoder-platform/resources-api
* https://github.com/topcoder-platform/challenge-api

The system will need to know that the user is in a group associated with a challenge, and be able to get the challenge ID for the group, to be able to call the resources and / or challenge APIs to check roles.

We can do this with just text next to a handle first, unless flair would be easier.  Something like `Ghostar (copilot)` or `TonyJ (admin)`

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30143397 has been created for this ticket.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by atelomycterus** [comment]
@jmgasper 
### Approach 1

a group has challengeId and we can use challenge/resource APIs and get roles before displaying pages in Vanilla.

### Approach 2
 
Using challenge processor forums + add a role in addUserToGroup + add the 'topcoder role' column in UserGroup.
No requests from Vanilla. 

topic: challenge.notification.create
  "payload": {
    "id": "58a85d42-4b1f-4fd1-a448-60c611a97677",
    "challengeId": "d28f85e8-1fa1-43d1-8fee-a44119ccee55",
    "memberId": "8547899",
    "memberHandle": "TonyJ",
    "roleId": "67ad906d-a63a-4337-b9cc-0a311d2600ad"
  }

Use resource API (/resource-roles) to get a name of the roleId.

 
topic: challenge.notification.events
  "payload": {
    "type": "USER_REGISTRATION",
    "data": {
      "challengeId": "d28f85e8-1fa1-43d1-8fee-a44119ccee55",
      "userId": 151743
    }
  }
  
Topcoder role is  'member'.

topic: challenge.notification.create 
we get a payload and can use "createdBy": "handle". Is the user an admin for this challenge? 
is it possible if user can have one more  roles in one challenge?

**Comment by jmgasper** [comment]
@atelomycterus - Let's go with approach 1.  There will be multiple users with admin and copilot roles per challenge, and the list of admins and copilots can change after the challenge has been created.

Here's an example of a recent challenge, with a bunch of managers and observers added by the system:

![Screen Shot 2020-09-25 at 09 16 39](https://user-images.githubusercontent.com/21790/94209529-dd437780-ff0f-11ea-97d1-9584b7c9251c.png)

**Comment by atelomycterus** [comment]
@jmgasper Almost implemented. Should I added roles to all links 'Started by XXX ' and so on for all group pages (/group/XXXX)? 

![image](https://user-images.githubusercontent.com/10196632/94738377-ff0d8600-0377-11eb-8694-728cfc6c17ba.png)

![image](https://user-images.githubusercontent.com/10196632/94738649-6592a400-0378-11eb-8dd5-bfa9cb9768bd.png)

**Comment by jmgasper** [comment]
@atelomycterus - Yeah, let's do it everywhere we can, unless it's difficult.

**Comment by atelomycterus** [comment]
@jmgasper Done. Please apply PRs:
	https://github.com/topcoder-platform/forums/pull/55
	https://github.com/topcoder-platform/forums-plugins/pull/3
	https://github.com/topcoder-platform/forums-groups-plugin/pull/8

Topcoder plugin's settings:
![image](https://user-images.githubusercontent.com/10196632/94785926-031dc000-03d9-11eb-8179-a8fbf133ac81.png)

![image](https://user-images.githubusercontent.com/10196632/94786325-7d4e4480-03d9-11eb-9d1e-af212ebf2e3c.png)

![image](https://user-images.githubusercontent.com/10196632/94786233-5ee84900-03d9-11eb-9910-39773f435827.png)

If a discussion has GroupId and Group with ChallengeID:
![image](https://user-images.githubusercontent.com/10196632/94786072-2ba5ba00-03d9-11eb-92fc-46cdf0f9b486.png)

**Comment by jmgasper** [comment]
@atelomycterus - I'm testing this but am not seeing it working.  Is there an easy way in the UI to investigate if the group has the challenge ID set?  I don't have access to the database...

**Comment by jmgasper** [comment]
![Screen Shot 2020-10-02 at 11 43 56](https://user-images.githubusercontent.com/21790/94879770-8eb05300-04a4-11eb-8948-e3b6ec428c06.png)

**Comment by atelomycterus** [comment]
@jmgasper if the ChallengeID is set then 'Challenge' label is added:
![image](https://user-images.githubusercontent.com/10196632/94888733-af67b100-0482-11eb-8a68-abddea2bc788.png)

'Challenge Overview'
'Justin Forum Test 11' is the challenge link in a discussion body. I think we need to add 'Challenge' link into the green box if it exist. 
![image](https://user-images.githubusercontent.com/10196632/94888771-c3abae00-0482-11eb-9015-b699b60ddee1.png)

**Comment by atelomycterus** [comment]
@jmgasper  M2M token configuration is invalid. I see the values of all params with a debug Bar on vanilla.topcoder-dev.com.

Tested with the same settings on my local env. It didn't work.  The color of 'TonyJ link' was green, it should be 'orange'.
Topcoder roles for a challenge were missing as well.

The 'Plugins.Topcoder.M2M.Auth0Audience' value is "https://api.topcoder.com/" but it should be 'https://m2m.topcoder-dev.com/'. 

I created PR https://github.com/topcoder-platform/forums/pull/58 and the value is hard-coded in config:
```
saveToConfig('Plugins.Topcoder.M2M.Auth0Audience','https://m2m.topcoder-dev.com/'); // getenv('AUTH0_AUDIENCE')
```
